specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -176,7 +222,47 @@ Field | Value | Description | Source
 `context.message.queue.name` | e.g. `my-topic` | The SNS topic name. | The 6th segment of `record.sns.topicArn`
 `context.message.age.ms` | e.g. `3298` | Age of the message in milliseconds. `current_time` - `snsTimestamp`.  | `record.sns.timestamp`
 `context.message.body` | - | The message body. Should only be captured if body capturing is enabled in the configuration. | `record.sns.message`
-`context.message.headers` | - | The message attributes. Should only be captured, if capturing headers is enabled in the configuration. | `record.sns.messageAttributes`
+`context.message.headers` | - | The message attributes. Should only be captured, if capturing headers is enabled in the configuration. | Use the `Value` of entries in `record.Sns.MessageAttributes` with `Type == "String"`. [Other attribute types](https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes) are ignored. |
```

Comment:
I think we should also capture the values of type `String.Array` and `Number` for SNS. But explicitly exclude `Binary` types.

Reply:
I'll see if I can get a `String.Array` to work... at least I couldn't initially from the *CLI*. Probably can with an aws-sdk library call.

From the link above:

> String.Array – An array, formatted as a string, that can contain multiple values. The values can be strings, numbers, or the keywords true, false, and null.

Those boolean and null values in the array are *not* allowed by our APM Server intake spec, I think.

Reply:
> formatted as a string

So those are always strings, right? Even if it's a boolean the value will be `'true'` as String.
Can't we just take the string value of the attribute as is?

Reply:
The spec also says this about `String.Array`:

> This data type isn't supported for AWS Lambda subscriptions. If you specify this data type for Lambda endpoints, it's passed as the String data type in the JSON payload that Amazon SNS delivers to Lambda. 

And when I publish an SNS message with these message attributes:

```
% cat send-message.json
{
  "City": {
    "DataType": "String",
    "StringValue": "Any City"
  },
  "Things": {
    "DataType": "String.Array",
    "StringValue": "[\"one\",\"two\"]"
  },
  "Greeting": {
    "DataType": "Binary",
    "BinaryValue": "Hello, World!"
  },
  "Population": {
    "DataType": "Number",
    "StringValue": "1250800"
  }
}

% aws sns publish \
    --topic-arn arn:aws:sns:us-west-2:627286350134:trentm-play-topic1 \
    --message 'this is my message' --subject 'this is my subject' \
    --message-attributes file://send-message.json
{
    "MessageId": "9783ee98-8ffd-5221-9d73-b43cb002fac9"
}
```

This is the event that I get in my Lambda:

```
...
{
  Greeting: { Type: 'Binary', Value: 'SGVsbG8sIFdvcmxkIQ==' },
  Things: { Type: 'String', Value: '["one","two"]' },
  Population: { Type: 'String', Value: '1250800' },
  City: { Type: 'String', Value: 'Any City' }
}
...
```

Notice that my inputs with `DataType` `"String.Array"` and `"Number"` both end up with `Type: 'String'`.

So... unless we can get a test/example case where we see a Lambda event with a `Type` other than 'String' or 'Binary', I think we should keep the allowlist to `Type == 'String'`.  Thoughts?